### PR TITLE
Polygon for DNS resolution

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/22147
Resolves https://github.com/brave/brave-browser/issues/22151

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

These domains could be used to test polygon DNS resolution
`uns-devtest-fda7a9.x` ([transaction](https://polygonscan.com/tx/0x0f779e1118f1c589b28e613d378ebe96b6cce95d772910210145f6251a249746))
`uns-devtest-brave.x`([transaction](https://polygonscan.com/tx/0x0f779e1118f1c589b28e613d378ebe96b6cce95d772910210145f6251a249746))
both should redirect to https://brave.com/

### CODE REVIEW

File: browser/net/decentralized_dns_network_delegate_helper_unittest.cc
Comment (kdenhartog) [inline browser/net/decentralized_dns_network_delegate_helper_unittest.cc:141]: Would we expect that a different protocol scheme would ever be used with unstoppable domains. Can we add some tests for those cases if so

**Review by kdenhartog — COMMENTED** [review]
Few comments for the sec team, but I didn't see a security review issue opened for this yet, but it looks like we're code owners on some of the changed files.

File: net/decentralized_dns/constants.h
Comment (yrliou) [reply net/decentralized_dns/constants.h:12]: We don't support Zilliqa Name Service (ZNS) at the moment, only Unstoppable Name Service (UNS).

File: browser/net/decentralized_dns_network_delegate_helper_unittest.cc
Comment (supermassive) [reply browser/net/decentralized_dns_network_delegate_helper_unittest.cc:141]: Purpose of this test is to cover these checks:
https://github.com/brave/brave-core/blob/04732b0cb16448c3d46fbabcd1ba1daec87d55d1/browser/net/decentralized_dns_network_delegate_helper.cc#L37
https://github.com/brave/brave-core/blob/e6227e93eb7723b9ea51a7a19be0cc212cda4a69/components/decentralized_dns/utils.cc#L38
So this is not about protocols support, but about proper tlds.

File: components/brave_wallet/browser/json_rpc_service_test_utils.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service_test_utils.cc:13]: Yes, it is technically possible to include test utils file into production code, but this should be blocked as something strange by PR reviewer in `#include` section and in gn deps section.

File: components/brave_wallet/browser/json_rpc_service_test_utils_unittest.cc
Comment (supermassive) [reply components/brave_wallet/browser/json_rpc_service_test_utils_unittest.cc:78]: Added test for an invalid url here
https://github.com/brave/brave-core/pull/13202/commits/bda3d8f597668a89de8b6b851fa966b9d59c7123

File: components/brave_wallet/browser/brave_wallet_utils.cc
Comment (yrliou) [reply components/brave_wallet/browser/brave_wallet_utils.cc:249]: @supermassive For easier to track in the future, could you put a comment in #22151 about which merged PR already covers the UD part and add this issue in this PR description to auto resolve this when merge? Thanks.

File: components/brave_wallet/browser/unstoppable_domains_multichain_calls.h
Comment (yrliou) [inline components/brave_wallet/browser/unstoppable_domains_multichain_calls.h:79]: nit: keep the newline

**Review by yrliou — APPROVED** [review]

File: chromium_src/net/base/lookup_string_in_fixed_set.cc
Comment (diracdeltas) [inline chromium_src/net/base/lookup_string_in_fixed_set.cc:49]: nit: should be `unstoppable_domain` not `unstoppabe_domain`

**Comment by diracdeltas** [comment]
does https://github.com/brave/brave-browser/wiki/Resolve-Methods-for-Unstoppable-Domains need an update here?

also could you remind me what is the expected behavior in Tor windows?

**Comment by supermassive** [comment]
> does https://github.com/brave/brave-browser/wiki/Resolve-Methods-for-Unstoppable-Domains need an update here?

There are some issues in parallel regarding deprecation of DNS over HTTPS UD resolver:
https://github.com/brave/brave-browser/issues/22148
https://github.com/brave/brave-browser/issues/22149
Was going to update wiki as a part of this
https://github.com/brave/brave-core/pull/13256#issuecomment-1122375252

> also could you remind me what is the expected behavior in Tor windows?

UD and ENS resolvers are disabled for incognito browsers:
https://github.com/brave/brave-core/blob/04732b0cb16448c3d46fbabcd1ba1daec87d55d1/browser/net/decentralized_dns_network_delegate_helper.cc#L27

File: chromium_src/net/base/lookup_string_in_fixed_set.cc
Comment (supermassive) [reply chromium_src/net/base/lookup_string_in_fixed_set.cc]: https://github.com/brave/brave-core/pull/13202/commits/4f916563f379813df4f0438dbd77bad3f10553d2

File: components/brave_wallet/browser/unstoppable_domains_multichain_calls.h
Comment (supermassive) [reply components/brave_wallet/browser/unstoppable_domains_multichain_calls.h:80]: https://github.com/brave/brave-core/pull/13202/commits/4f916563f379813df4f0438dbd77bad3f10553d2

File: components/brave_wallet/browser/unstoppable_domains_dns_resolve.cc
Comment (supermassive) [reply components/brave_wallet/browser/unstoppable_domains_dns_resolve.cc:63]: https://github.com/brave/brave-core/pull/13202/commits/4f916563f379813df4f0438dbd77bad3f10553d2

File: components/brave_wallet/browser/brave_wallet_utils.cc
Comment (supermassive) [reply components/brave_wallet/browser/brave_wallet_utils.cc]: https://github.com/brave/brave-core/pull/13202/commits/4f916563f379813df4f0438dbd77bad3f10553d2

File: components/brave_wallet/browser/brave_wallet_utils.cc
Comment (supermassive) [reply components/brave_wallet/browser/brave_wallet_utils.cc:249]: done

File: components/brave_wallet/browser/json_rpc_service.cc
Comment (kdenhartog) [reply components/brave_wallet/browser/json_rpc_service.cc:1258]: Ok ideally we'd want to use a 200 here unless there was an expected reason to handle another 200 class HTTP response. This is to have more defensive programming logic in case the Polygon gateway changes underneath us. This isn't a blocking change though and I'll leave it up to you whether or not you think it's necessary to change.

File: components/brave_wallet/browser/brave_wallet_constants.h
Comment (iefremov) [inline components/brave_wallet/browser/brave_wallet_constants.h:625]: constexpr

File: chromium_src/net/dns/dns_transaction.cc
Comment (iefremov) [inline chromium_src/net/dns/dns_transaction.cc:22]: `base::ranges::any_of` !

**Review by iefremov — APPROVED** [review]

File: chromium_src/net/dns/dns_transaction.cc
Comment (supermassive) [reply chromium_src/net/dns/dns_transaction.cc]: https://github.com/brave/brave-core/pull/13202/commits/7d207e078107633244e3172548633c6c7187aff1

File: components/brave_wallet/browser/brave_wallet_constants.h
Comment (supermassive) [reply components/brave_wallet/browser/brave_wallet_constants.h]: added many of them
https://github.com/brave/brave-core/pull/13202/commits/7d207e078107633244e3172548633c6c7187aff1

File: components/brave_wallet/browser/unstoppable_domains_dns_resolve.cc
Comment (goodov) [inline components/brave_wallet/browser/unstoppable_domains_dns_resolve.cc:20]: `kStyled` constants are preferred around Chromium. see also `brave_wallet_types.h`.
https://google.github.io/styleguide/cppguide.html#Enumerator_Names

**Review by goodov — APPROVED** [review]
chromium_src lgtm, but some other parts can be improved.

File: components/brave_wallet/browser/unstoppable_domains_dns_resolve.cc
Comment (supermassive) [reply components/brave_wallet/browser/unstoppable_domains_dns_resolve.cc:13]: kept plain enum so there are no static_casts when used as index in vector

File: components/brave_wallet/browser/unstoppable_domains_dns_resolve.cc
Comment (goodov) [inline components/brave_wallet/browser/unstoppable_domains_dns_resolve.cc:28]: huh, I was expecting that it will become a one row thing :)

Comment (supermassive) [reply components/brave_wallet/browser/unstoppable_domains_dns_resolve.cc:28]: It seems it doesn't work like that :D

File: browser/net/decentralized_dns_network_delegate_helper.cc
Comment (iefremov) [reply browser/net/decentralized_dns_network_delegate_helper.cc:84]: actually it has been with us for so long that I'm thinking of keeping it as a memorial
(just kidding)

**Review by kdenhartog — APPROVED** [review]
Thanks, my concerns have been addressed. Looks good to me

File: browser/net/decentralized_dns_network_delegate_helper_unittest.cc
Comment (bridiver) [inline browser/net/decentralized_dns_network_delegate_helper_unittest.cc:51]: This does not make sense in a unit test, but this test is not a unit test anyway so it should be converted to a browser test. Unit tests are for a "unit" of code like a class or single file

**Pull request merged**
